accord-core/src/main/java/accord/impl/progresslog/WaitingState.java

```diff
@@ -0,0 +1,624 @@
+            owner.clearActive(Waiting, txnId);
+            set(safeStore, owner, blockedUntil, Queued);
+        }
+    }
+
+    void record(DefaultProgressLog owner, SaveStatus newSaveStatus)
+    {
+        BlockedUntil currentlyBlockedUntil = blockedUntil();
+        if (currentlyBlockedUntil.minSaveStatus.compareTo(newSaveStatus) <= 0)
+        {
+            boolean isDone = newSaveStatus.hasBeen(Status.PreApplied);
+            set(null, owner, isDone ? CanApply : currentlyBlockedUntil, NoneExpected);
+            if (isDone)
+                maybeRemove(owner);
+            owner.clearActive(Waiting, txnId);
+        }
+    }
+
+    final void runWaiting(SafeCommandStore safeStore, SafeCommand safeCommand, DefaultProgressLog owner)
+    {
+        run(safeStore, safeCommand, owner);
+    }
+
+    private void run(SafeCommandStore safeStore, SafeCommand safeCommand, DefaultProgressLog owner)
+    {
+        BlockedUntil blockedUntil = blockedUntil();
+        Command command = safeCommand.current();
+        Invariants.checkState(!owner.hasActive(Waiting, txnId));
+        Invariants.checkState(command.saveStatus().compareTo(blockedUntil.minSaveStatus) < 0, "Command has met desired criteria but progress log entry has not been cancelled");
+
+        set(safeStore, owner, blockedUntil, Querying);
+        TxnId txnId = safeCommand.txnId();
+        // first make sure we have enough information to obtain the command locally
+        Timestamp executeAt = command.executeAtIfKnown();
+        EpochSupplier toLocalEpoch = toLocalEpoch(safeStore, txnId, blockedUntil, command, executeAt);
+        Participants<?> fetchKeys = Invariants.nonNull(command.maxContactable());
+
+        if (!Route.isRoute(fetchKeys))
+        {
+            fetchRoute(owner, blockedUntil, txnId, executeAt, toLocalEpoch, fetchKeys);
+            return;
+        }
+
+        Route<?> route = Route.castToRoute(fetchKeys);
+        if (homeSatisfies().compareTo(blockedUntil) < 0)
+        {
+            awaitHomeKey(owner, blockedUntil, txnId, executeAt, route);
+            return;
+        }
+
+        Route<?> slicedRoute = slicedRoute(safeStore, txnId, route, toLocalEpoch);
+        if (!command.hasBeen(Status.PreCommitted))
+        {
+            // we know it has been decided one way or the other by the home shard at least, so we attempt a fetch
+            // including the home shard to get us to at least PreCommitted where we can safely wait on individual shards
+            fetch(owner, blockedUntil, txnId, executeAt, toLocalEpoch, slicedRoute, slicedRoute.withHomeKey());
+            return;
+        }
+
+        Route<?> awaitRoute = awaitRoute(slicedRoute, blockedUntil);
+        Route<?> fetchRoute = fetchRoute(slicedRoute, awaitRoute, blockedUntil);
+
+        if (slicedRoute.size() == 0 || awaitRoute.isHomeKeyOnlyRoute())
+        {
+            // at this point we can switch to polling as we know someone has the relevant state
```

Comment:
because the home shard can satisfy our query, and we are waiting either on the home shard only, or we don't even intersect the command's keys so need some universally-applicable state like executeAt